runtime/jcl/common/thread.cpp

```diff
@@ -603,28 +603,110 @@ Java_java_lang_Thread_setExtentLocalCache(JNIEnv *env, jclass unusedClass, jobje
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
 #if JAVA_SPEC_VERSION >= 20
+/**
+ * Frame walk function to find the most recent scoped value binding in the stack.
+ *
+ * @param currentThread the current thread
+ * @param walkState the stack walk state
+ * @return J9_STACKWALK_KEEP_ITERATING or J9_STACKWALK_STOP_ITERATING
+ */
+static UDATA
+findScopedValueBindingsWalkFunction(J9VMThread *currentThread, J9StackWalkState *walkState)
+{
+	UDATA rc = J9_STACKWALK_KEEP_ITERATING;
+	if (NULL != walkState->method) {
+		J9JavaVM *vm = currentThread->javaVM;
+		J9Method *method = walkState->method;
+		J9Class *methodClass = J9_CLASS_FROM_METHOD(method);
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+		J9UTF8 *methodName = J9ROMMETHOD_NAME(romMethod);
+
+		if (((J9VMJAVALANGTHREAD_OR_NULL(vm) == methodClass)
+			|| (J9VMJAVALANGVIRTUALTHREAD_OR_NULL(vm) == methodClass)
+			|| (J9VMJAVALANGSCOPEDVALUECARRIER_OR_NULL(vm) == methodClass))
+			&& J9UTF8_LITERAL_EQUALS(J9UTF8_DATA(methodName), J9UTF8_LENGTH(methodName), "runWith")
+		) {
+			if (NULL != walkState->userData1) {
+				rc = J9_STACKWALK_STOP_ITERATING;
+			} else {
+				Assert_JCL_true(0);
+			}
+		} else if (0 == walkState->inlineDepth) {
+			walkState->userData1 = NULL;
+		}
+	}
+
+	return rc;
+}
+
+/**
+ * Object slot walk function to find the most recent scoped value binding in the object slots.
+ *
+ * @param currentThread the current thread
+ * @param walkState the stack walk state
+ * @param slot pointer to slot containing an object pointer
+ * @param stackLocation pointer to the slot in the stack containing possibly compressed object pointer
+ */
+static void
+findScopedValueBindingOSlotWalkFunction(J9VMThread *currentThread, J9StackWalkState *walkState, j9object_t *slot, const void *stackLocation)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	j9object_t slotObject = *slot;
+	J9Class *clazz = J9OBJECT_CLAZZ(currentThread, slotObject);
+	J9Class *snapshotClazz = J9VMJAVALANGSCOPEDVALUESNAPSHOT_OR_NULL(vm);
```

Comment:
This is invalid - the class may not be loaded, which will cause a crash.

Reply:
Should `J9VMJAVALANGSCOPEDVALUESNAPSHOT` be used instead?

Reply:
I think that results in calling `findKnownClass` if the entry is not resolved. I think this is fine here, but it will need to be done while holding VM access (i.e. after entering the VM).

Reply:
Is this the same concern as mentioned in https://github.com/eclipse-openj9/openj9/pull/17402#discussion_r1294484802? I've added one field of `ScopedValue$Snapshot` to the constant pool.